bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+            },
+        };
+        let key_pattern = match sub_matches.value_of(KEY_ARG) {
+            None => None,
+            Some(p) => match Self::parse_key_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // again
+            },
+        };
+
+        Self {
+            subcommand,
+            file_pattern,
+            key_pattern,
+            keysize: sub_matches.value_or_default(KEY_SIZE_ARG),
+            host: sub_matches.value_or_default(HOST_ARG),
+            port: sub_matches.value_or_default(PORT_ARG),
+        }
+    }
+    fn parse_file_pattern(file_pattern: &str) -> Result<FilePattern<'_>, ParseError> {
+        let path = PathBuf::from(file_pattern);
+        let name = path.file_name().unwrap().to_str().unwrap();
+
+        let re = Regex::new("\\{key\\}").unwrap();
+        let two_or_more = Regex::new(".*\\{key\\}.*\\{key\\}.*").unwrap();
+
+        if re.is_match(name) {
+            if two_or_more.is_match(name) {
+                Err(ParseError::FilePattern)
+            } else {
+                let path = PathBuf::from(file_pattern);
+
+                let dir = match path.parent().unwrap() {
+                    p if p == Path::new("") => Path::new("."),
+                    p => p,
+                };
+                Ok(FilePattern::WithRE(
+                    RePattern::new(file_pattern),
+                    dir.to_owned(),
+                ))
+            }
+        } else {
+            Ok(FilePattern::WithoutRE(file_pattern))
+        }
+    }
+    fn parse_key_pattern(key_arg_cont: &str) -> Result<KeyPattern, ParseError> {
+        if let Ok(key) = key_arg_cont.parse() {
+            return Ok(KeyPattern::Single(key));
+        }
+        if key_arg_cont.contains("-") {
+            let v: Vec<&str> = key_arg_cont.split("-").collect();
+            if v.len() != 2 || v[0] >= v[1] {
+                return Err(ParseError::KeyPattern);
+            }
+            let v: Vec<u64> = v.into_iter().map(|n| n.parse().unwrap()).collect();
+            let range = v[0]..=v[1];
+            return Ok(KeyPattern::Range(range));
+        }
+        Ok(KeyPattern::Multiple(
+            key_arg_cont
+                .split(",")
+                .map(|n| n.parse().unwrap())
+                .collect(),
+        ))
+    }
+}
+
+#[derive(Debug)]
+enum FilePattern<'a> {
+    WithRE(RePattern<'a>, PathBuf),
+    WithoutRE(&'a str),
+}
+
+#[derive(Debug)]
+struct RePattern<'a> {
+    inner: &'a str,
+}
+
+impl<'a> RePattern<'a> {
+    fn new(f: &'a str) -> Self {
+        Self { inner: f }
+    }
+
+    fn get_regex(&self) -> Regex {
+        lazy_static! {
+            static ref RE: Regex = Regex::new("\\\\\\{key\\\\\\}").unwrap();
+        }
+        let st = RE
+            .replace(&regex::escape(self.inner), "(\\d+)")
+            .into_owned();
+        Regex::new(&st).unwrap()
+    }
+    fn get_filenames(&self, keys: &KeyPattern) -> Vec<(u64, String)> {
```

Comment:
It is better to return an iterator here, because `KeyPattern::Range` can produce a lot of keys and it is better to process them lazily rather than collect them into a vector